### ISSUE TITLE
Wasm: Locals Exceed Maxmimum

### ISSUE THREAD

**Issue by theduke** [RustPython/RustPython#3054]
I've tried to run RustPython in wasm via `wasmer` and `wasmtime`.

Both give me an error when trying to use this code built with `--target=wasm32-wasi`:

```
Error: failed to run main module `target/wasm32-wasi/debug/semantic_plugin_python.wasm`

Caused by:
    0: WebAssembly failed to compile
    1: WebAssembly translation error
    2: Invalid input WebAssembly code at offset 24846689: locals exceed maximum
```

```rust
use rustpython_vm as vm;
use rustpython_vm::{InitParameter, Interpreter, PySettings, VirtualMachine};

pub fn main() {

    let mut scope = None;
    let interp = vm::Interpreter::new_with_init(PySettings::default(), |vm| {
            // VM_INIT_FUNCS.with(|cell| {
            //     for f in cell.borrow().iter() {
            //         f(vm)
            //     }
            // });
            vm.add_frozen(rustpython_derive::py_freeze!(dir = "pyenv/lib/python3.9/site-packages/bs4"));

            scope = Some(vm.new_scope_with_builtins());

            InitParameter::Internal
        });

    let scope = scope.unwrap();

    interp.enter(|vm| -> Result<(), Box<dyn std::error::Error>> {
        // let scope = vm.new_scope_with_builtins();


        let code_obj = vm
            .compile(
                r#"print("Hello World!")"#,
                vm::compile::Mode::Exec,
                "<embedded>".to_owned(),
            )
            .map_err(|err| vm.new_syntax_error(&err)).unwrap();

        vm.run_code_obj(code_obj, scope).unwrap();

        Ok(())
    }).unwrap();
}
```

This seems to be a WASM validation error, so maybe this is a `rustc` codegen issue?

**Comment by matklad** [comment]
Yeah, facing the same issue. I *think* we are hitting the following limit here: 

https://github.com/bytecodealliance/wasm-tools/blob/ac9028a4c52dbf1b0e78b24be33b7e5ce9618ca6/crates/wasmparser/src/limits.rs#L29

This is not a spec limit, its an implementation limit.


I, however, only hit this in debug mode. If I compile in release, then I hit some runtime error.

**Comment by bhelx** [comment]
I was also able to confirm the release build works, but also got this to stop this by bumping the opt-level for the dev profile to `1`:

```
[profile.dev]
opt-level = 1
```

But problem probably needs to be addressed long term.

I searched the resulting wasm code and I think this function is the culprit:

```
rustpython_parser::python::__parse__Top::__reduce
```

**Comment by DimitrisJim** [comment]
> I think this function is the culprit:

Probably. I remember seeing this function mentioned elsewhere for something wasm related. Don't know if we can do much to reduce the size of that function.

**Comment by youknowone** [comment]
Maybe lalrpop can inline less in the function

**Comment by joshuamegnauth54** [comment]
I know this issue is a bit old but I recently encountered _and_ fixed it after an hour of research. To save others the time, especially since this issue specifically is one of the top hits, I'll summarize what I found.

The base issue seems to be present across WebAssembly runtimes. The spec doesn't define a limit on locals, but it seems like runtimes all sort of gravitated towards the same limit. Our C++ friends have had similar problems:

* emscripten-core/emscripten#19346
* emscripten-core/emscripten#21847

Browser wasm virtual machines seem to have the same limit too. I'm legitimately unsure if this information is outdated by now. 🤔 

To fix it, I copied the profiles from `RustPython`'s base workspace `Cargo.toml`. It's a simple solution and one hinted at above, but it took a while to figure out.

From `RustPython`:

```toml
[profile.dev.package."*"]
opt-level = 3

[profile.test]
opt-level = 3
# https://github.com/rust-lang/rust/issues/92869
# lto = "thin"

[profile.bench]
lto = "thin"
codegen-units = 1
opt-level = 3

[profile.release]
lto = "thin"
```

My issue involved [wasm_bindgen_test](https://docs.rs/wasm-bindgen-test/latest/wasm_bindgen_test/) failing to run tests due to exceeding the locals limit. The test profile changes fixed it.

---

Finally, as a small extra, I also encountered an issue with `getrandom`, a dependency of `RustPython`.  Weirdly enough, `rustpython_wasm`, the crate, fails to compile unless the wasm backend for `getrandom` is explicitly chosen. The RustPython repo builds because it picks the backend, but using the `rustpython_wasm` crate fails because cargo doesn't use the backend when building the crate for some reason. To fix that, I created a `.cargo` directory with a `config.toml` manually selecting the backend:

```toml
[target.wasm32-unknown-unknown]
rustflags = ["--cfg", 'getrandom_backend="wasm_js"']
```

And I also had to manually include `getrandom` as a dependency with the correct feature enabled:

```toml
[dependencies]
getrandom = { version = "0.3", default-features = false, features = [
  "wasm_js",
] }
```
I'm not using `getrandom` anywhere but I had to include it anyway or else the build failed.

The above behavior is surprising since I would expect it to just work. In any case, I hope this helps anyone trawling through search engines in need of a solution.

**Comment by youknowone** [comment]
@joshuamegnauth54 Thank you!
